### PR TITLE
GH-1636 add support for semantic actions

### LINKED ISSUE #1636: Add support for Semantic Actions
### Version

main

### Feature

The ShEx test suite has [tests](https://github.com/shexSpec/shexTest/blob/6afa7ef08144d1d7fa45845012cede12b881506e/validation/manifest.ttl#L1086-L1106) for [semantic actions](http://shex.io/shex-semantics/index.html#semantic-actions). Maybe some kindly soul will add support for them.

### Are you interested in contributing a solution yourself?

Yes

### PR DESCRIPTION
**Pull request by ericprud**
GitHub issue resolved #1636 

Pull request Description:

Adds support or all SemAct traits except ExternalSemanticAction. That will require invoking another ShExC parser and configuring the validator with pointers to the semantic actions parsed from the auxiliary ShEx test file.

----

 - [x] Tests are included.
 - [ ] Documentation change and updates are provided for the [Apache Jena website](https://github.com/apache/jena-site/)
 - [x] Commits have been squashed to remove intermediate development commit messages.
 - [x] Key commit messages start with the issue number (GH-xxxx or JENA-xxxx)

By submitting this pull request, I acknowledge that I am making a contribution to the Apache Software Foundation under the terms and conditions of the [Contributor's Agreement](https://www.apache.org/licenses/contributor-agreements.html).

----

See the [Apache Jena "Contributing" guide](https://github.com/apache/jena/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java [hunk 1]
```diff
@@ -632,6 +641,48 @@ protected void numericFacetLength(String facetKind, int length, int line, int co
...
 
+    protected SemAct crackSemanticAction(String iriAndCode, int line, int column) {
+        // e.g. % <http://shex.io/extensions/Test/> { print(s) %}
+        // or   % ex:Test { print(s) %}
+        // or   %<http://shex.io/extensions/Test/>%
+
+        // Trim leading '%' and ws and trim trailing '%}'
+        String whitespaces = " \t\n\r\f";
+        int startOfIri = 1; // get past '%'
+        for (; whitespaces.indexOf(iriAndCode.charAt(startOfIri)) != -1; ++startOfIri)
```
Comment (rvesse) [inline jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java:652]: This all seems a bit brittle e.g. what if there's excess whitespace anywhere?

We have a commons-lang dependency already so `StringUtils.stripStart()` would be more robust

File: jena-shex/src/main/java/org/apache/jena/shex/parser/javacc/ShExJavacc.java [hunk 2]
```diff
@@ -1,4 +1,3 @@
-/* ShExJavacc.java */
```
Comment (rvesse) [inline jena-shex/src/main/java/org/apache/jena/shex/parser/javacc/ShExJavacc.java:1]: Again unclear how much of the diff here is due to using an older JavaCC version vs actual functional changes from the grammar changes

File: jena-shex/src/main/java/org/apache/jena/shex/sys/ValidationContext.java [hunk 3]
```diff
@@ -79,37 +127,81 @@ public void startValidate(ShexShape shape, Node data) {
...
+    public boolean dispatchStartSemanticAction(ShexSchema schema, ValidationContext vCxt) {
+        return !schema.getSemActs().stream().anyMatch(semAct -> {
+            String semActIri = semAct.getIri();
+            SemanticActionPlugin semActPlugin = this.semActPluginIndex.get(semActIri);
+            if (semActPlugin != null) {
+                if (!semActPlugin.evaluateStart(semAct, schema)) {
+                    vCxt.reportEntry(new ReportItem(String.format("%s start shape failed", semActIri), null));
+                    return true;
+                }
+            }
```
Comment (rvesse) [inline jena-shex/src/main/java/org/apache/jena/shex/sys/ValidationContext.java:142]: It would be helpful to report that a semantic action plugin was unavailable so users have some indication of why validation failed

File: jena-shex/src/test/files/spec/schemas/1dotCode3fail.shex [hunk 4]
```diff
@@ -1,6 +1,11 @@
+PREFIX p: <http://shex.io/extensions/>
```
Comment (rvesse) [inline jena-shex/src/test/files/spec/schemas/1dotCode3fail.shex:1]: Per my comments elsewhere it would be useful to have a variety of test cases here with variations in whitespace, use of URIs vs prefixed names to identify plugins etc, to check that the parser modifications are sufficiently robust

Although looking at other changes perhaps the reduction in test exclusions is already picking up a selection of pre-existing test cases that were previously skipped?

File: jena-shex/src/test/java/org/apache/jena/shex/runner/ShexT.java [hunk 5]
```diff
@@ -74,4 +81,12 @@ public class ShexT {
     public final static Property sx_json = ResourceFactory.createProperty(NS_SX + "json");
     public final static Property sx_ttl  = ResourceFactory.createProperty(NS_SX + "ttl");
 
+
+    /** <p>The test statusThe expected output of semantic actions</p> */
```
Comment (rvesse) [inline jena-shex/src/test/java/org/apache/jena/shex/runner/ShexT.java:85]: Some copy-paste leftovers here?

File: jena-shex/src/test/java/org/apache/jena/shex/runner/ShexTests.java [hunk 6]
```diff
@@ -329,10 +325,41 @@ private static boolean runTestExclusionsInclusions(ManifestEntry entry) {
                 //System.err.println("Skipping:  "+fragment);
                 return false;
             }
+
+            List<Resource> traits = extractTraits(entry);
+            if (traits != null) {
+                List<Resource> excludedBecause = traits.stream().filter(excl -> excludeTraits.contains(excl)).collect(Collectors.toList());
```
Comment (rvesse) [inline jena-shex/src/test/java/org/apache/jena/shex/runner/ShexTests.java:331]: Think you can simplify the whole interior of this if block to just `return !traits.stream().anyMatch(excl -> excludeTraits.contains(excl));`

**Review by rvesse — CHANGES_REQUESTED** [review]
Thanks for the contribution, I'm not a ShEx expert so I'll let someone else review the substantive feature work

There are a few comments mostly niggles

The main substantive one that does need addressing is that you appear to have used an older JavaCC to regenerate some of the parsers which is making the PR very noisy in places.  Using JavaCC 7 to regenerate those parsers as was previously used by the project would hopefully reduce noise in this PR substantially and make it easier to focus on the actual feature enhancement

**Comment by ericprud** [comment]
@rvesse , Tx for detailed review! I'm going to cook up more SemAct use cases for ShEx to test whitespace variants. ETA 2hrs on specific responses and fixes.

**Comment by afs** [comment]
@ericprud -- this needs to ready in days, rather than hours.

Experience shows, big changes benefit from a period of stability before release (even better - wider testing).

There are changes elsewhere which impact many users.

Status:
https://lists.apache.org/thread/29fbynbom80jtyf7nl2lchh0lj93cwq9

**Comment by ericprud** [comment]
@afs, since you dev on linux, did you download a non-deb version of javacc? I only have
```
javacc/jammy,jammy,now 5.0-8.1 all
```

**Comment by afs** [comment]
Ubuntu (kinetic/22.10) : `javacc/kinetic,kinetic 7.0.11-1 all`

**Comment by ericprud** [comment]
> Ubuntu (kinetic/22.10) : `javacc/kinetic,kinetic 7.0.11-1 all`

Tx. I'd downloaded a javacc tarball which brought be back to a paleolithic 4.1. I grabbed 7.0.11 from [ubuntu](https://packages.ubuntu.com/kinetic/all/javacc/download) and see "7.0" in (some of) the generated files.

I've pushed the generated parser files without a rebase to make comments easier to track. Next up, whitespace-y tests to demonstrate that [`crackSemanticAction`](https://github.com/apache/jena/pull/1637/files#diff-a7ad2334b126ad2223fdabe8a053448b814ff5fbe2a77fc29e0eb754140ff942R644) gracefully handles everything the parser accepts.

File: jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java (hunk 1 above)
Comment (ericprud) [reply jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java:652]: I totally agree that it looks like 80's era K&R C, but I engineered this one pretty hard and I don't think that StringUtils will add anything but cycles. I [added tests](https://github.com/apache/jena/pull/1637/files#diff-f1b4d58d2a4e81046c490afc8b71496e8440f6aed8e15c3ded996bdb8de0c4c9) to verify that whatever the parser accepted, my C ptrs could handle.

File: jena-shex/src/main/java/org/apache/jena/shex/sys/ValidationContext.java (hunk 3 above)
Comment (ericprud) [reply jena-shex/src/main/java/org/apache/jena/shex/sys/ValidationContext.java:142]: Indeed. How about a contract where a plugin can [return (and passively register) a specific URL](https://github.com/apache/jena/pull/1637/files#diff-eac8e39fb9aaa5feae63ecad83a13992b4caef8b84ecc757239a7e4382875630R45-R49) which gets called for any unregistered SemAct? You could pass an implementation that throws or logs or pages a programmer.

File: jena-shex/src/test/files/spec/schemas/1dotCode3fail.shex (hunk 4 above)
Comment (ericprud) [reply jena-shex/src/test/files/spec/schemas/1dotCode3fail.shex:1]: Oops, I was messing with that while I was fiddling with the parser. Restored old state with presumption that the aforementioned [SemAct tests](https://github.com/apache/jena/pull/1637/files#diff-f1b4d58d2a4e81046c490afc8b71496e8440f6aed8e15c3ded996bdb8de0c4c9) provide better coverage.

File: jena-shex/src/test/java/org/apache/jena/shex/runner/ShexT.java (hunk 5 above)
Comment (ericprud) [reply jena-shex/src/test/java/org/apache/jena/shex/runner/ShexT.java]: Typo faithfully copied from `jena-core/src/main/java/org/apache/jena/vocabulary/TestManifest.java:42`
fixed.

File: jena-shex/src/test/java/org/apache/jena/shex/runner/ShexTests.java (hunk 6 above)
Comment (ericprud) [reply jena-shex/src/test/java/org/apache/jena/shex/runner/ShexTests.java:331]: I compose the `excludedBecause` because (because because) I used it to see why tests were excluded. It could be nice to have a command-line arg to log it but at least this way you can println it to chase down mysteries.

File: jena-shex/src/main/java/org/apache/jena/shex/expressions/ShapeExprAtom.java [hunk 7]
```diff
@@ -31,6 +31,7 @@ public class ShapeExprAtom extends ShapeExpression {
     private final ShapeExpression other;
 
     public ShapeExprAtom(ShapeExpression other) {
+        super(null);
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/expressions/ShapeExprAtom.java:34]: These super(null) are unnecessary. Add a protected constructor `ShapeExpression() { this(null); }`

File: jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java (hunk 1 above)
Comment (afs) [reply jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java:652]: The code is for everyone to read ... including the future-eric who has rid himself of the 1980's.

File: jena-shex/src/main/java/org/apache/jena/shex/parser/javacc/ParseException.java [hunk 8]
```diff
@@ -83,7 +78,7 @@ public ParseException(String message) {
   /**
    * This is the last token that has been consumed successfully.  If
    * this object has been created due to a parse error, the token
-   * following this token will (therefore) be the first error token.
+   * followng this token will (therefore) be the first error token.
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/parser/javacc/ParseException.java]: typo - "following"

File: jena-shex/src/main/java/org/apache/jena/shex/eval/ShapeEvalEachOf.java [hunk 9]
```diff
@@ -97,7 +97,7 @@ else if ( partitions.isEmpty() )
             }
             // This partition works.
             if ( success )
-                return true;
+                return eachOf.testSemanticActions(vCxt, matchables);
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/eval/ShapeEvalEachOf.java:100]: Correct the comment on L99.

File: jena-shex/src/main/java/org/apache/jena/shex/expressions/TripleExpression.java [hunk 10]
```diff
@@ -18,16 +18,43 @@
...
+        return semActs;
+    }
+
+    public void setSemActs(List<SemAct> semActs) { // needed for ShExC parser's late binding of SemActs to EachOf
+        this.semActs = semActs;
+    }
 
-    protected TripleExpression() {}
+//    public void setSemActs(List<SemAct> semActs) {
+//        this.semActs = semActs;
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/expressions/TripleExpression.java:50]: Remove

File: jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java (hunk 1 above)
Comment (afs) [reply jena-shex/src/main/java/org/apache/jena/shex/parser/ParserShExC.java:652]: It adds clarity.

File: jena-shex/src/main/java/org/apache/jena/shex/semact/TestSemanticActionPlugin.java [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+import java.util.regex.Pattern;
+
+interface ExtractVar {
+    String run (String str);
+}
+
+public class TestSemanticActionPlugin implements SemanticActionPlugin {
+    static String SemActIri = "http://shex.io/extensions/Test/";
+    static Pattern ParsePattern = Pattern.compile("^ *(fail|print) *\\((\\\"(?:(?:[^\\\\\\\"]|\\\\[^\\\"])+)\\\"|[spo])\\) *$");
+    //       tic Pattern ParsePattern = Pattern.compile("^ *(fail|print) *\\((\\\"(?:(?:[^\\\\\\\"]|\\\\[^\\\"])+)\\\"|[spo])\\) *$");
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/semact/TestSemanticActionPlugin.java:41]: Remove unused

File: jena-shex/src/main/java/org/apache/jena/shex/Plugin.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.jena.shex;
+
+import org.apache.jena.shex.sys.SysShex;
+
+import java.util.List;
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/Plugin.java:23]: Remove unused imports.

There are about 10 of them across the PR.

File: jena-shex/src/main/java/org/apache/jena/shex/sys/SysShex.java [hunk 13]
```diff
@@ -36,13 +43,33 @@ public class SysShex {
...
         return systemValiditor;
     }
+
+    static {
+        semActPluginIndex = new ConcurrentHashMap<>();
+        systemValiditor = new ShexValidatorImpl(semActPluginIndex);
+    }
+
+    public static ShexValidator getNew(Collection<SemanticActionPlugin> pz) {
+        Map iriToPlugin = new ConcurrentHashMap();
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/sys/SysShex.java:68]: `Map` needs generics. `ConcurrentHashMap` needs `<>`

File: jena-shex/src/main/java/org/apache/jena/shex/sys/ValidationContext.java [hunk 14]
```diff
@@ -18,45 +18,93 @@
...
         // Fresh ShexReport.Builder
-        return new ValidationContext(vCxt.data, vCxt.shapes, vCxt.inProgress);
+        return new ValidationContext(vCxt, vCxt.data, vCxt.shapes, vCxt.inProgress, vCxt.semActPluginIndex);
+    }
+
+    /**
+     * Creates a new validation context with the current one as its parent context.
+     * Initializes the new context with the state of the parent context.
+     *
+     * @return
```
Comment (afs) [inline jena-shex/src/main/java/org/apache/jena/shex/sys/ValidationContext.java:59]: Remove or add the type because `@return` on its own is a javadoc warning.

**Review by afs — CHANGES_REQUESTED** [review]
Partial code-level review.

**Review by afs — APPROVED** [review]
